### PR TITLE
AddNetwork: Avalanche Fuji Testnet Support

### PR DESCRIPTION
**Pull request by kamescg**
Adding support for the Avalanche Fuji 🗻🗻🗻 testnet.

### CODE REVIEW

File: helpers/handleBeaconChainContractDeploy.ts [hunk 1]
```diff
@@ -0,0 +1,7 @@
+import { Signer } from "@ethersproject/abstract-signer";
+
+export async function handleBeaconChainContractDeploy(deploy: Function, deployer: string) {
+
+}
+
+export default handleBeaconChainContractDeploy
```
Comment (asselstine) [inline helpers/handleBeaconChainContractDeploy.ts:7]: ✂️  this file, looks like it.  No placeholders!

File: helpers/handlePeripheryContractDeploy.ts [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+  reserve: ''
+}) {
+  const prizeFlushResult = await deployContract(deploy, 'PrizeFlush', deployer, [deployer, config.prizeDistributor, config.prizeSplitStrategy, config.reserve])
+
+  return {
+    prizeFlushResult: prizeFlushResult,
+  }
+}
+
+export default handlePeripheryContractDeploy
```
Comment (asselstine) [inline helpers/handlePeripheryContractDeploy.ts:21]: hmm....

`handlePeripheryContractDeploy` is only called in one place.  This file adds *a lot* of code without any real benefit other than consistency in approach.  Less code is better!  Basically a maxim.

✂️

File: helpers/handlePrizePoolCoreDeploy.ts [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+export async function handlePrizePoolCoreDeploy(
+  deploy: Function,
+  deployer: string,
+  ethers: any,
+  config: handlePrizePoolCoreDeployConfig = {
+    decimals: 18,
+    yieldSourcePrizePool: '',
+    drawDufferCardinality: "",
+    prizeDistributionBufferCardinality: ""
+  }
```
Comment (asselstine) [inline helpers/handlePrizePoolCoreDeploy.ts:21]: I think named params here would be simpler than creating a separate typescript type.  Less overhead.

File: helpers/handleReceiverChainContractDeploy.ts [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+export async function handleReceiverChainContractDeploy(deploy: Function, deployer: string, config: handleReceiverChainContractDeployConfig = {
+  drawCalculator: '',
+  drawBuffer: '',
+  prizeDistributionBuffer: ''
+}) {
+  await deployContract(deploy, 'L2TimelockTrigger', deployer, [deployer, config.drawBuffer, config.prizeDistributionBuffer, config.drawCalculator])
+}
+
+export default handleReceiverChainContractDeploy
```
Comment (asselstine) [inline helpers/handleReceiverChainContractDeploy.ts:18]: again...this file is only called in one place.  Just put the function there!  Doesn't need a file + type + separate declaration

**Review by asselstine — CHANGES_REQUESTED** [review]
Looks pretty good, although I'm seeing lots of boilerplate code for the sake of consistency.  It's like `createBigNumber` but on a larger scale :)

**Less is more**

Only refactor and create helper functions *if you need to*!

File: helpers/handleBeaconChainContractDeploy.ts (hunk 1 above)
Comment (kamescg) [reply helpers/handleBeaconChainContractDeploy.ts:7]: Fair point, but also trying to understand how I want to organize everything. Still need to reactor the Rinkeby function into that "placeholder".

File: helpers/handlePeripheryContractDeploy.ts (hunk 2 above)
Comment (kamescg) [reply helpers/handlePeripheryContractDeploy.ts:21]: I would argue consistency with approach is pretty important. It's weird to switch in the middle of execution.

File: helpers/handleBeaconChainContractDeploy.ts (hunk 1 above)
Comment (kamescg) [reply helpers/handleBeaconChainContractDeploy.ts:41]: Placeholder filled 😏

File: helpers/handleReceiverChainContractDeploy.ts (hunk 4 above)
Comment (kamescg) [reply helpers/handleReceiverChainContractDeploy.ts:18]: It's called in the mumbai deploy script now too.

File: helpers/handlePrizePoolCoreDeploy.ts (hunk 3 above)
Comment (kamescg) [reply helpers/handlePrizePoolCoreDeploy.ts:21]: Personally as a developer I wouldn't want to worry about ordering of contract addresses, and instead rely on named parameters for a more simple setup.

Any other opinions @ddesrosier @PierrickGT @aodhgan?

File: helpers/handlePeripheryContractDeploy.ts (hunk 2 above)
Comment (asselstine) [reply helpers/handlePeripheryContractDeploy.ts:24]: You've added an extra file and 21 lines of code for no additional benefit.

This is way too much overhead.

File: helpers/handlePrizePoolCoreDeploy.ts (hunk 3 above)
Comment (asselstine) [reply helpers/handlePrizePoolCoreDeploy.ts:21]: Using **named params** is better.  It's not about ordering.

What you have above is that some of the params are ordered, and the final param is essential a named param.

Having named params for the whole function is a much simpler interface.

Comment (asselstine) [reply helpers/handlePrizePoolCoreDeploy.ts:21]: Here's a [quick explainer](https://blog.jcoglan.com/2021/06/07/named-arguments-in-javascript/)

although I'm guessing you're already familiar with it, maybe just don't know the name

File: helpers/handleBeaconChainContractDeploy.ts (hunk 1 above)
Comment (asselstine) [reply helpers/handleBeaconChainContractDeploy.ts:41]: This file is still empty....?  Should be deleted.  Dead code

**Review by asselstine — CHANGES_REQUESTED** [review]

**Comment by kamescg** [comment]
@asselstine The new PR adds code to that file. So not empty.

https://github.com/pooltogether/v4-testnet/blob/pool-1901-v4-testnet-add-avalanche-testnet/helpers/handleBeaconChainContractDeploy.ts

**Comment by kamescg** [comment]
@asselstine just bumped the beta package version https://www.npmjs.com/package/@pooltogether/v4-testnet/v/1.0.0-beta.14 but before merging into master perhaps we can do a quick overview of the structure?

To get ready for mainnet deployment started migrating updated deployment scripts to v4-mainnet. So if we're happy with this setup we can start focusing on the custom upgrade scripts for existing PrizePools in the PrizePoolNetwork too.

https://github.com/pooltogether/v4-mainnet/pull/4

The current PR has PrizeDistributionFactory, BeaconTimelockAndPushRouter, ReceiverTimelockAndPushRouter, and TwabRewards contracts included.

**Pull request merged**